### PR TITLE
Add state and integrate model builder

### PR DESCRIPTION
**Pull request by maanug-nv**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do ?

Replaces Megatron's module-level training globals with an explicit `GlobalState`, and wires the config-driven model builder into the `pretrain` setup path (closer to how `megatron-bridge` does it).

- **`GlobalState` / `TrainState`** (new `state.py`): single object holding `cfg`, `tokenizer`, `timers`, `train_state`, signal handler, straggler timer, and async checkpoint queue. `TrainState` is a `Stateful` dataclass for checkpointing.
- **Model builder integration**: `setup_model_and_optimizer` builds via the config's model builder when given a `cfg_container`/`pg_collection`, falling back to legacy `get_model` otherwise. `model_provider` is now optional and moved to the end of `pretrain`'s signature; entrypoints updated accordingly.
- **Setup helpers** extracted into new `setup.py` (`maybe_save_config`, `init_checkpointing_context`, `validate_and_set_vocab_size`); `pretrain` reads from the config container / state instead of `args.*` in many places.
- **Supporting**: typing + logging cleanup in `dist_signal_handler.py`; `barrier_and_log` absorbs `print_datetime`; new `train_utils.start_memory_history_recording`; `--enable-experimental` moved into `DistributedInitConfig`; added `CheckpointConfig.async_write_results_mp_mode`.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

**Comment by maanug-nv** [comment]
seems that some TE bumps recently caused numerics drift for the MoE op-fuser test, so I had to update the golden values. 

cc @Connor-XY 
> I've looked into this and I believe the drift is overwhelmingly the TE revision bump (#4992/#5112), very plausibly compounded by the #4859 grouped-MLP-fuser refactor on the SwiGLU path. I believe you could update golden values to fix the current test

**Review by tdene — APPROVED** [review]

File: megatron/training/utils/utils.py
Comment (jaredcasper) [inline megatron/training/utils/utils.py:1]: It's already in "training/utils", why not just "utils.py"?

File: tests/functional_tests/test_cases/moe/gpt3_mcore_te_tp1_pp1_te_4experts_groupedGEMM_op_fuser/golden_values_dev_dgx_h100.json
Comment (jaredcasper) [inline tests/functional_tests/test_cases/moe/gpt3_mcore_te_tp1_pp1_te_4experts_groupedGEMM_op_fuser/golden_values_dev_dgx_h100.json:1]: Why do the golden values need to change?

File: megatron/training/utils/utils.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+from megatron.core._rank_utils import safe_get_rank
+from megatron.training.config import ProfilingConfig
+
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+def start_memory_history_recording(profiling: ProfilingConfig | None) -> None:
```
Comment (jaredcasper) [inline megatron/training/utils/utils.py:15]: This seems like an orthogonal change to the state and model builder stuff?

File: tests/functional_tests/test_cases/moe/gpt3_mcore_te_tp1_pp1_te_4experts_groupedGEMM_op_fuser/golden_values_dev_dgx_h100.json
Comment (jaredcasper) [reply tests/functional_tests/test_cases/moe/gpt3_mcore_te_tp1_pp1_te_4experts_groupedGEMM_op_fuser/golden_values_dev_dgx_h100.json:1]: oh, just saw comment above, this should be separate.

File: megatron/training/utils/utils.py (hunk 1 above)
Comment (maanug-nv) [reply megatron/training/utils/utils.py:15]: it's mirroring what's happening in model_provider.py : https://github.com/NVIDIA/Megatron-LM/blob/main/model_provider.py. After this PR, that will no longer be in use by pretrain scripts.

File: megatron/training/utils/utils.py
Comment (maanug-nv) [reply megatron/training/utils/utils.py:1]: done

File: megatron/training/config/common_config.py [hunk 2]
```diff
@@ -151,3 +151,6 @@ class DistributedInitConfig:
 
     disable_jit_fuser: bool = False
     """Disable the JIT fuser."""
+
+    enable_megatron_core_experimental: bool = field(default=False, metadata={"argparse_meta": {"arg_names": ["--enable-experimental"]}})
```
Comment (jaredcasper) [inline megatron/training/config/common_config.py:155]: why do we need "megatron_core" here? Why not just "enable_experimental"? Are there experimental features outside of core that this doesn't impact or something?

File: megatron/training/dist_signal_handler.py [hunk 3]
```diff
@@ -1,78 +1,134 @@
...
+    def __enter__(self) -> "DistributedSignalHandler":
+        from megatron.training.utils import print_rank_0
+
         self._signal_received = False
         self.released = False
         self.original_handler = signal.getsignal(self.sig)
 
-        def handler(signum, frame):
+        def handler(signum: int, frame: Any | None) -> None:
+            print_rank_0(f"Received signal {signum}, initiating graceful stop")
```
Comment (jaredcasper) [inline megatron/training/dist_signal_handler.py:114]: How do we know every signal is going to initiate a graceful stop?

File: megatron/training/state.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+from megatron.core.tokenizers.utils.build_tokenizer import build_tokenizer
+from torch.distributed.checkpoint.stateful import Stateful
+from torch.utils.tensorboard.writer import SummaryWriter
+
+from megatron.training.config import PretrainConfigContainer
+from megatron.training.dist_signal_handler import DistributedSignalHandler
+
+
+@dataclass
+class TrainState(Stateful):
```
Comment (jaredcasper) [inline megatron/training/state.py:24]: So is the usage of this class coming in a separate PR?

File: megatron/training/training.py [hunk 5]
```diff
@@ -1077,19 +1075,24 @@ def pretrain(
...
     timestamp_after_initialize_megatron = time.time()
 
     args = get_args()
     timers = get_timers()
+    # Inject the legacy timers directly; GlobalState.timers intentionally exposes
+    # no public setter (overriding it is not supported behavior). Temporary.
+    state._timers = timers
 
-    if args.fine_grained_activation_offloading:
+    if cfg_container.model.fine_grained_activation_offloading:
```
Comment (jaredcasper) [inline megatron/training/training.py:1089]: I know this is from an old change, but having `_container` seems verbose and clunky, why not just "cfg"?

File: megatron/training/dist_signal_handler.py (hunk 3 above)
Comment (maanug-nv) [reply megatron/training/dist_signal_handler.py:114]: this is just a message to indicate that the signal handler caught the configured signal to catch. iirc, after a signal (say sigterm) is sent, the training loop will exit at the end of the current step. 
i guess this probably doesn't work if a sigkill is sent. 

can you suggest a better message? or are you suggesting to just remove 'initiating graceful stop'
